File: src/feishu_auth_kit/probe.py
from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from .client import FeishuAuthClient

AI_AGENT_PING_PATH = "/open-apis/bot/v1/openclaw_bot/ping"


@dataclass(frozen=True)
class FeishuProbeResult:
    ok: bool
    app_id: str | None = None
    bot_name: str | None = None
    bot_open_id: str | None = None
    error: str | None = None
    raw: dict[str, Any] = field(default_factory=dict)


def register_ai_agent(client: FeishuAuthClient) -> FeishuProbeResult:
    """Validate credentials and register the app as an official Feishu/Lark AI agent."""
    try:
        tenant_token = client.get_tenant_access_token().token
        payload = client._request_json(
            "POST",
            f"{client.domains.open_base}{AI_AGENT_PING_PATH}",
            headers={"Authorization": f"Bearer {tenant_token}"},
            json={"needBotInfo": True},
        )
    except Exception as exc:
        return FeishuProbeResult(ok=False, app_id=client.app_id, error=str(exc))

    bot_info = payload.get("data", {}).get("pingBotInfo") or {}
    return FeishuProbeResult(
        ok=True,
        app_id=client.app_id,
        bot_name=bot_info.get("botName"),
        bot_open_id=bot_info.get("botID"),
        raw=payload,
    )


def probe_ai_agent_credentials(
    app_id: str,
    app_secret: str,
    *,
    brand: str = "feishu",
    session: Any | None = None,
) -> FeishuProbeResult:
    client = FeishuAuthClient(app_id, app_secret, brand=brand, session=session)
    return register_ai_agent(client)

File: src/probe.ts
import { FeishuAuthClient } from "./client.js";

export const AI_AGENT_PING_PATH = "/open-apis/bot/v1/openclaw_bot/ping";

export interface FeishuProbeResult {
  ok: boolean;
  app_id?: string | null;
  bot_name?: string | null;
  bot_open_id?: string | null;
  error?: string | null;
  raw?: Record<string, any>;
}

export async function registerAiAgent(client: FeishuAuthClient): Promise<FeishuProbeResult> {
  try {
    const tenantToken = (await client.getTenantAccessToken()).token;
    const url = `${client.domains.openBase}${AI_AGENT_PING_PATH}`;

    const response = await client.fetchFn(url, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${tenantToken}`,
        "Content-Type": "application/json",
        Accept: "application/json",
      },
      body: JSON.stringify({ needBotInfo: true }),
      signal: AbortSignal.timeout(client.timeoutMs),
    });

    if (!response.ok) {
      const errText = await response.text().catch(() => "");
      return {
        ok: false,
        app_id: client.appId,
        error: `HTTP ${response.status}: ${errText}`,
      };
    }

    const payload = await response.json();
    if (payload?.code !== 0 && payload?.code !== undefined && payload?.code !== null) {
      return {
        ok: false,
        app_id: client.appId,
        error: payload.msg || payload.message || "Unknown error",
        raw: payload,
      };
    }

    const botInfo = payload?.data?.pingBotInfo || {};
    return {
      ok: true,
      app_id: client.appId,
      bot_name: botInfo.botName ?? null,
      bot_open_id: botInfo.botID ?? null,
      raw: payload,
    };
  } catch (exc: any) {
    return {
      ok: false,
      app_id: client.appId,
      error: exc?.message || String(exc),
    };
  }
}

export async function probeAiAgentCredentials(
  appId: string,
  appSecret: string,
  options?: {
    brand?: string;
    fetchFn?: any;
  }
): Promise<FeishuProbeResult> {
  const client = new FeishuAuthClient(appId, appSecret, options);
  return registerAiAgent(client);
}
